snapcraft/internal/build_providers/_snap.py

```diff
@@ -0,0 +1,296 @@
+                "Unhandled scenario for {!r} (host installed: {}, latest_revision {})".format(
+                    self.snap_name, snap_repo.installed, self._latest_revision
+                )
+            )
+
+        self.__required_operation = op
+        return op
+
+    def get_assertions(self) -> List[List[str]]:
+        op = self.get_op()
+        if op != op.INJECT:
+            return []
+        snap_repo = self._get_snap_repo()
+        snap_info = snap_repo.get_local_snap_info()
+
+        if snap_info["revision"].startswith("x"):
+            return []
+
+        assertions = []  # type: List[List[str]]
+        assertions.append(["snap-declaration", "snap-name={}".format(snap_repo.name)])
+        assertions.append(
+            [
+                "snap-revision",
+                "snap-revision={}".format(snap_info["revision"]),
+                "snap-id={}".format(snap_info["id"]),
+            ]
+        )
+        return assertions
+
+    def _set_data(self) -> None:
+        op = self.get_op()
+        snap_repo = self._get_snap_repo()
+        install_cmd = ["sudo", "snap"]
+
+        if op == _SnapOp.INJECT:
+            install_cmd.append("install")
+            snap_info = snap_repo.get_local_snap_info()
+            snap_revision = snap_info["revision"]
+
+            if snap_revision.startswith("x"):
+                install_cmd.append("--dangerous")
+
+            if snap_info["confinement"] == "classic":
+                install_cmd.append("--classic")
+
+            snap_file_name = "{}_{}.snap".format(snap_repo.name, snap_revision)
+            install_cmd.append(os.path.join(self._snap_dir, snap_file_name))
+        elif op == _SnapOp.INSTALL or op == _SnapOp.REFRESH:
+            install_cmd.append(op.name.lower())
+            snap_info = snap_repo.get_store_snap_info()
+            snap_revision = snap_info["channels"]["latest/stable"]["revision"]
+            confinement = snap_info["channels"]["latest/stable"]["confinement"]
+            if confinement == "classic":
+                install_cmd.append("--classic")
+            install_cmd.append(snap_repo.name)
+        elif op == _SnapOp.NOP:
+            install_cmd = []
+            snap_revision = None
+
+        self.__install_cmd = install_cmd
+        self.__revision = snap_revision
+
+    def get_revision(self) -> str:
+        if self.__revision is None:
+            self._set_data()
+        return self.__revision
+
+    def get_install_cmd(self) -> List[str]:
+        if self.__install_cmd is None:
+            self._set_data()
+        return self.__install_cmd
+
+
+_STORE_ASSERTION = [
+    "account-key",
+    "public-key-sha3-384=BWDEoaqyr25nF5SNCvEv2v7QnM9QsfCc0PBMYD_i2NGSQ32EF2d4D0hqUel3m8ul",
+]
+
+
+class SnapInjector:
+    def __init__(
+        self,
+        *,
+        snap_dir: str,
+        registry_filepath: str,
+        runner,
+        snap_dir_mounter,
+        snap_dir_unmounter,
+        file_pusher
```

Comment:
Would like to see some types specified for these callbacks.

Reply:
yes, I forgot!

Reply:
there is no comprehensive way to define a Callable that takes optional arguments nor one that takes kwargs https://docs.python.org/3.5/library/typing.html#callable